src/Conversion/ONNXToMhlo/Tensor/Concat.cpp

```diff
@@ -0,0 +1,64 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===---------------- Concat.cpp - Lowering Concat Op -------------------===//
+//
+// Copyright 2019-2022 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX Concat Operator to Mhlo dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToMhlo/ONNXToMhloCommon.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+struct ONNXConcatOpLoweringToMhlo : public ConversionPattern {
+  ONNXConcatOpLoweringToMhlo(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXConcatOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+
+    Location loc = op->getLoc();
+    ONNXConcatOpAdaptor operandAdaptor(operands);
+    ONNXConcatOp concatOp = llvm::cast<ONNXConcatOp>(op);
+
+    if (op->getNumResults() < 1) {
+      op->emitError() << "ONNXConcatOp Has No Output\n";
+      return failure();
+    }
+    RankedTensorType resultType =
+        op->getResult(0).getType().dyn_cast_or_null<RankedTensorType>();
+    if (resultType == nullptr) {
+      op->emitError() << "Concat Output Is Not Ranked\n";
+      return failure();
```

Comment:
Not specific to this PR, I am a bit concerned about failures during rewrite patterns that can cause the whole compilation pipeline terminated. This would work if all ONNX ops can be lowered to MHLO. However, if an op does not satisfy the condition here, there is no a fallback path so that the op will be lowered to other dialects such as Krnl for CPU or ZHigh for NNPA.

One way to solve this is to use `DynamicallyLegalOp` in `--convert--onnx-to-mhlo` pass to selectively add operations into the conversion target. We did this for `--convert-onnx-to-zhigh`: https://github.com/onnx/onnx-mlir/blob/main/src/Accelerators/NNPA/Conversion/ONNXToZHigh/ONNXToZHigh.cpp#L323

FYI @AlexandreEichenberger

Reply:
Thanks for the comment! Will discuss with co-workers about the fallback path for mhlo lowering. @yaochengji

Reply:
@tungld  Currently `-convert-onnx-to-hlo` uses `applyPartialConversion` and it doesn't make ONNX Dialect or any ONNX operations illegal. Therefore it is fine if there're some ONNX ops couldn't be converted to MHLO.

Reply:
@yaochengji I see. Thanks for clarifying that!